File: 백준/Silver/9012. 괄호/괄호.py
import sys

input = sys.stdin.readline

n = int(input())
for _ in range(n):
    ps = list(input().rstrip())

    stack = []
    for i in ps:
        if i == '(':
            stack.append(i)
        elif i == ')' and len(stack) > 0:
            stack.pop()
        elif len(stack) == 0 and i == ')':
            stack.append(i)
            break

    if len(stack) == 0:
        print("YES")
    else:
        print("NO")

File: 백준/Silver/9012. 괄호/괄호.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Scanner;
import java.util.Stack;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int count = Integer.parseInt(br.readLine());

        for (int i = 0; i < count; i++) {
            String ex = br.readLine();
            if (isVPS(ex)) {
                System.out.println("YES");
            } else {
                System.out.println("NO");
            }
        }
    }

    public static boolean isVPS(String ex) {
        Stack<String> stack = new Stack<>();

        try {
            for (int i = 0; i < ex.length(); i++) {
                if (ex.charAt(i) == '(') {
                    stack.push("(");
                } else if (ex.charAt(i) == ')') {
                    stack.pop();
                }
            }

            return stack.isEmpty();
        } catch (Exception e){
            return false;
        }
    }
}